refactor: replace `moment-timezone` with `Intl`-based timezone utilities

KristjanESPERANTO opened the pull request:
- replace the parser’s `moment-timezone` dependency with a lightweight `Intl`-powered `tz-utils` layer that keeps TZID metadata, Windows alias handling, and offset parsing intact  
  - With that, we now have correct all-day event timezone handling. All-day events retain their original timezone instead of being normalized to UTC midnight. This prevents wrong calendar days for non-UTC events and DST transitions. Consumers should use the attached timezone for local day calculations. This makes node-ical more reliable and standards-compliant.
- add unit tests for the new helper plus snapshot coverage that runs all recurrence examples under a fixed zone to guard against regressions
- refresh docs and examples to demonstrate RRULE expansion with Moment, Luxon, Day.js, date-fns, and vanilla Intl while keeping moment-timezone as an examples-only dev dependency

I tried again to make the commits as comprehensible as possible. It's still a big load, though.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added multiple RRULE example scripts demonstrating recurrence expansion, overrides, exceptions, and formatted instance output across several date libraries and a vanilla JS example.

- **Documentation**
  - Overhauled README with explicit RRULE and timezone guidance, “Working with the parsed dates”, Intl and Temporal examples, migration notes, and links to synchronous/RRULE examples.

- **Tests**
  - Added snapshot tests for examples, comprehensive tz-utils unit tests, and a metadata-preservation test for all-day DTSTART.

- **Chores**
  - Replaced runtime Moment usage with an Intl-backed tz-utils, updated dev dependencies and example runner.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

KristjanESPERANTO committed:
refactor(tz): introduce `tz-utils` layer and decouple parser from `moment-timezone`

- add `tz-utils.js` with helpers:
  `guessLocalZone`, `getZoneNames`, `findExactZoneMatch`, `isValidIana`,
  `parseDateTimeInZone`, `parseWithOffset`, `utcAdd`, `formatMMMMDoYYYY`, `linkAlias`
- refactor `ical.js` to use `tz-utils` (no direct `moment-timezone` imports)
- tests:
  - remove `moment-timezone` from test suite
  - add unit tests `test/tz-utils.test.js`
  - replace moment-specific checks in `test/basic.test.js` and `test/advanced.test.js` with `tz-utils` and stable `Date`-based assertions
  - rename suite titles for clarity (“unit: tz-utils”, “integration: fromURL”)

Validation:
- all tests green (51)

Notes:
- no functional changes; prepares for later TZ library swap

KristjanESPERANTO committed:
refactor(tz): migrate timezone logic from moment-timezone to Intl-based implementation

`moment-timezone` is not completely removed from `package.json` so that the example js files continue to work -but it moves from `dependencies` to `devDependencies`.

KristjanESPERANTO committed:
test: add snapshot tests for all examples scripts

KristjanESPERANTO committed:
feat(examples): add multiple examples for expanding recurring events using various date libraries

KristjanESPERANTO committed:
fix(tz-utils): handle edge case for hour value of 24 in parseDateTimeInZone function

KristjanESPERANTO committed:
fix(tests): set timezone in snapshot tests

KristjanESPERANTO committed:
fix(tests): normalize newlines in example output for snapshot comparison

KristjanESPERANTO commented:
I want to check a few things, so I've set the PR to draft for now.

But maybe you'd still like to take a look and see if the direction I've taken is okay for you :slightly_smiling_face:

KristjanESPERANTO committed:
refactor(tz): refactor timezone handling to use shared helpers

- replace bespoke Windows/offset detection in `ical.js` with `tz-utils.resolveTZID`
- emit RRULE DTSTART strings via `formatDateForRrule`, falling back when metadata is incomplete
- export the new helpers from `tz-utils` so callers can share the timezone logic

KristjanESPERANTO committed:
test: document synthetic minute-offset tzid case

KristjanESPERANTO committed:
refactor(tz): rebuild all-day DTSTART at local midnight

- replace string-based midnight fallback with direct Y/M/D reconstruction in `ical.js`
- drop the unused `formatMMMMDoYYYY` helper from `tz-utils.js`

KristjanESPERANTO committed:
test: cover DST transitions in tz-utils

KristjanESPERANTO committed:
refactor(tests): replace toLocaleString with toLocaleDateString for date formatting

KristjanESPERANTO committed:
refactor(tz): normalize TZID parsing across parameters

- read TZID from every parameter entry, update the normalized label in-place, and reuse the extracted date parts for conversions
- improves calendar accuracy for end users by preventing missed timezones in unusual ICS files
- share the existing stack-based timezone fallback to keep the parsing logic consistent

KristjanESPERANTO committed:
examples: dedupe recurring RRULE output across variants

- key RRULE expansions by calendar day so each occurrence prints once
- merge RECURRENCE-ID overrides into the map before rendering
- update all example-rrule-* demos to share the same override/dedupe flow

KristjanESPERANTO committed:
refactor(tests): enhance comments for clarity

KristjanESPERANTO committed:
refactor(tz): improve local zone guessing and zone name retrieval

- getZoneNames() now returns the complete catalog (Node 18+)
- remove legacy fallback to local zone
- update tests to assert local zone is present in the full list

KristjanESPERANTO committed:
test(todos): update assertion for completed task time to reflect UTC offset

KristjanESPERANTO committed:
test(tz): simplify and harden local timezone detection test

- Set TZ explicitly to a valid IANA zone (Europe/Berlin)
- Assert guessLocalZone returns the set zone
- Assert zone is present in IANA zone list
- Remove redundant assertions for clarity

KristjanESPERANTO commented on package.json:
```diff
@@ -21,13 +21,16 @@
     "rrule": "2.8.1"
   },
   "devDependencies": {
+    "date-fns": "^4.1.0",
+    "dayjs": "^1.11.18",
     "fast-xml-parser": "^5.2.5",
     "lint-staged": "^16.1.6",
+    "luxon": "^3.7.2",
     "mocha": "^11.7.2",
+    "moment-timezone": "^0.6.0",
     "simple-git-hooks": "^2.13.1",
```
This change is intentional: with this PR, node-ical no longer depends on moment-timezone or other date libraries, making the core package lighter and independent. Users can choose which date library to install for their own use case, or stick with vanilla JS as shown in the vanilla example.

This approach avoids unnecessary dependencies for users who don’t need them.

KristjanESPERANTO commented on tz-utils.js:
```diff
@@ -0,0 +1,266 @@
+}
+
+function getZoneNames() {
+  // Without shipping a full IANA db list, return a minimal array that satisfies tests
+  return [guessLocalZone()].filter(Boolean);
+}
```
The implementation has been updated, so this regression should be fully resolved.

KristjanESPERANTO commented:
@jens-maus I have reviewed the nitpicks and comments from CodeRabbit and made changes. The PR is now ready for me to remove the draft status :smiley: 

There are a lot of changes and a lot to review. I hope it is manageable and that you will find the time for it.

KristjanESPERANTO commented:
Okay, I'm checking the new nitpicks and comments :slightly_smiling_face:

KristjanESPERANTO committed:
fix: preserve metadata when normalizing all-day DTSTART

- keep tz/dateOnly flags after rebuilding RRULE DTSTART dates
- document the intent inline to highlight the non-obvious reattach step
- add a parser regression test covering metadata retention for all-day events

KristjanESPERANTO commented on examples/example-rrule-datefns.js:
```diff
@@ -0,0 +1,132 @@
+    if (isBefore(endDate, rangeStart) || isAfter(startDate, rangeEnd)) {
+      showRecurrence = false;
+    }
+
```
Keeping the current logic - deduping RRULE instances by calendar day—makes sense for this example because it’s meant to show a day-based agenda view. That’s why the code keeps only one occurrence per day, even if the rule has multiple events in the same 24-hour period.

KristjanESPERANTO committed:
examples: cover final day in RRULE ranges

KristjanESPERANTO committed:
examples: keep RRULE overrides while retaining duplicates

- key expanded occurrences by full ISO timestamp across vanilla/date-fns/moment/dayjs/luxon demos
- skip base RRULE instances when a RECURRENCE-ID override exists so replacements don’t double-print
- preserve day-based lookup keys for EXDATE/recurrence merging
- all snapshots stay in sync via `npm test`

KristjanESPERANTO committed:
examples: add minimal RRULE demo and cover it with docs/tests

KristjanESPERANTO committed:
chore(tz): document Node 18+ assumption in getZoneNames

KristjanESPERANTO commented on tz-utils.js:
```diff
@@ -0,0 +1,445 @@
+  return new Intl.DateTimeFormat().resolvedOptions().timeZone;
+}
+
+function getZoneNames() {
+  return Intl.supportedValuesOf('timeZone');
+}
```
We depend on Node 18+, so `Intl.supportedValuesOf('timeZone')` is guaranteed to exist and yields the canonical list without requiring extra guards.

KristjanESPERANTO commented:
This bunch is damn complex! I've now worked through the new nitpicks and comments from CodeRabbit.

KristjanESPERANTO commented on examples/example-rrule-vanilla.js:
```diff
@@ -0,0 +1,147 @@
+      console.log(`duration:${formatDuration(curDuration)}`);
+      console.log();
+    }
+  }
```
Done.

KristjanESPERANTO commented on examples/example-rrule-moment.js:
```diff
@@ -21,11 +21,13 @@ const ical = require('../node-ical.js');
 // Use a fixed date range to keep expansion finite (recurrences can be unbounded).
-const rangeStart = moment('2017-01-01');
-const rangeEnd = moment('2017-12-31');
+const rangeStart = moment('2017-01-01').startOf('day');
+const rangeEnd = moment('2017-12-31').endOf('day');
```
We already force a zone when running the examples via `examples-snapshot.test.js`. With that env set, startOf('day')/endOf('day') stay consistent across CI, so no additional pinning is needed.

KristjanESPERANTO commented on tz-utils.js:
```diff
@@ -0,0 +1,451 @@
+ * We depend on Node 18+, so `Intl.supportedValuesOf('timeZone')` is guaranteed
+ * to exist and yields the canonical list without requiring extra guards.
+ */
+function getZoneNames() {
+  return Intl.supportedValuesOf('timeZone');
+}
```
We’ve already got that covered - the main CI workflow (`nodejs.yml`) runs a matrix over node 18.x, 20.x, 22.x and 24.x on all three OS targets. So every job runs on Node 18 or newer, matching the engines field.

KristjanESPERANTO committed:
docs: review "Working with the parsed dates" section

KristjanESPERANTO committed:
refactor(examples): remove unnecessary callback in RRULE date expansion

KristjanESPERANTO committed:
fix(examples): handle potential null end dates in recurrence duration calculations

KristjanESPERANTO committed:
fix: resolve TZIDs before parsing

Ensuring aliased identifiers get canonical handling.

KristjanESPERANTO committed:
fix: enhance date parsing to support extended ISO format in `parseWithOffset`

KristjanESPERANTO committed:
refactor(examples): remove unused Day.js plugins from example script

KristjanESPERANTO committed:
fix(examples): ensure date range bounds are pinned to UTC for consistent date conversion

KristjanESPERANTO committed:
fix(docs): format code block for better readability in README

KristjanESPERANTO commented on README.md:
````diff
@@ -215,54 +218,49 @@ ical.fromURL('http://lanyrd.com/topics/nodejs/nodejs.ics', {}, function (err, da
-    }
-});
-```
+    If your runtime ships [Temporal](https://tc39.es/proposal-temporal/), you can also round-trip via:
+
+        ```js
+        const zoned = Temporal.ZonedDateTime.from({
+            timeZone: event.start.tz ?? 'UTC',
+            instant: Temporal.Instant.fromEpochMilliseconds(event.start.valueOf()),
+        }).startOfDay();
+        ```
+
````
Done.

KristjanESPERANTO commented on examples/example-rrule-luxon.js:
```diff
@@ -0,0 +1,140 @@
+
+// Use a fixed date range to keep expansion finite (recurrences can be unbounded).
+const rangeStart = DateTime.fromISO('2017-01-01').startOf('day');
+const rangeEnd = DateTime.fromISO('2017-12-31').endOf('day');
```
Done.

KristjanESPERANTO committed:
fix(tests): increase timeout for RRULE + EXDATE test to prevent spurious timeouts on Windows CI

KristjanESPERANTO committed:
fix: add validation for DURATION format to prevent errors during parsing

KristjanESPERANTO committed:
fix: normalize UTC/GMT prefixes when parsing offset labels

- strip optional “UTC/GMT” wrappers (with or without parentheses) in one pass
- drop a trailing “)” so inputs like “(UTC+02:00)” and “UTC+2” match the offset regex
- expands offsetLabelToMinutes to cover Windows-style TZ strings without changing other behaviour

KristjanESPERANTO committed:
refactor(tz): make Windows zone lookup deterministic

- removes substring heuristics so Windows labels resolve predictably to their canonical IANA zone
- precomputes normalized variants (case, spacing, offset prefixes, comma segments) for legacy display names
- keeps the parser resilient to messy inputs while avoiding accidental matches across similar Windows labels

KristjanESPERANTO committed:
style: keep tz metadata out of serialized output

- mark the injected tz flag as non-enumerable to avoid leaking metadata when events are stringified
- retain the internal access pattern without surprising downstream consumers

KristjanESPERANTO committed:
refactor(tz): extract DST convergence helper for `parseDateTimeInZone`

- add shared `ymdhmsToUtcMs` and `convergeLocalInstant` helpers so the fixed-point loop lives in one place
- increase the DST convergence loop to three passes and reuse it from `parseDateTimeInZone`
- makes the DST handling easier to reason about and keeps locale-specific tricks out of the parser core

KristjanESPERANTO committed:
docs(examples): make luxon recurrence keys use UTC dates

Luxon’s `toISO()` formatted the loop keys in the example using the
DateTime’s current zone, so running the script outside UTC could build
local-time keys and miss RECURRENCE-ID/EXDATE matches. Convert each
occurrence and recurrence ID to UTC before deriving the ISO string so
the keys stay deterministic across environments.

KristjanESPERANTO committed:
fix(tz): honor numeric offset when parsing VTIMEZONE fallback

The stack-based fallback that handles VTIMEZONE TZIDs only used
parseDateTimeInZone when an IANA name resolved, otherwise it fell back to
new Date(...). If resolveTZID yields a fixed numeric offset we now call
parseWithOffset so floating DTSTART values retain their wall time even
when the feed only publishes an offset-based definition.

KristjanESPERANTO committed:
fix(tz): keep Date.tz metadata consistently non-enumerable

Direct assignments in the parser (TZID parameters, UTC fallbacks,
recurrence overrides) were creating enumerable `tz` properties again.
Export the hardened helper from tz-utils and reuse it so we always
attach timezone metadata via Object.defineProperty, preventing JSON
leakage and keeping round-trips stable.

KristjanESPERANTO committed:
fix(tz): parse raw offset TZIDs in resolveTZID

resolveTZID previously only recognised offset metadata when the TZID was
wrapped in “(UTC±hh:mm)” style parentheses. Extend the parsing to reuse
offsetLabelToMinutes on bare strings like “UTC+02:00” or “+0530” so
feeds that expose fixed offsets without the Windows label still get
numeric offset metadata instead of falling back to local time semantics.

KristjanESPERANTO committed:
fix(tz): validate numeric offsets before parsing local wall times

parseWithOffset used to fall back to UTC when the offset regex failed,
masking bad input. Reuse offsetLabelToMinutes so malformed strings raise a
TypeError instead of silently mis-parsing, and add a unit test to document
the error path.

KristjanESPERANTO commented on tz-utils.js:
```diff
@@ -0,0 +1,455 @@
+  const minutesComponent = (oh * 60) + omm;
+  const totalMinutes = sign * minutesComponent;
+  const utcMs = Date.UTC(year, month - 1, day, hour, minute, second) - (totalMinutes * 60_000);
+  return new Date(utcMs);
```
Done with another approach: 5c5dd7c938aee1a0d00fec0b7078dfd602ae6b41.

KristjanESPERANTO commented on tz-utils.js:
```diff
@@ -0,0 +1,455 @@
+    tz = guessLocalZone();
+  }
+
+  tz = tz.replace(/^"(.*)"$/, '$1');
+  const original = tz;
+
+  if (tz && (tz.includes(' ') || tz.includes(','))) {
+    const mapped = mapWindowsZone(tz);
+    if (mapped) {
+      tz = mapped;
+    }
+  }
+
+  let offsetMinutes;
+  if (tz && tz.startsWith('(')) {
+    const offsetMatch = tz.match(/([+-]\d{1,2}:\d{2})/);
+    if (offsetMatch) {
+      offsetMinutes = offsetLabelToMinutes(offsetMatch[1]);
+    }
+
+    tz = null;
+  }
+
+  const exact = findExactZoneMatch(tz);
+  const iana = exact || (tz && isValidIana(tz) ? tz : undefined);
+  const offset = minutesToOffset(offsetMinutes);
+  const etc = minutesToEtcZone(offsetMinutes);
+
```
Done with 988ca76427614c798795befeccdaeea9335f846a.

KristjanESPERANTO commented on README.md:
````diff
@@ -215,54 +218,49 @@ ical.fromURL('http://lanyrd.com/topics/nodejs/nodejs.ics', {}, function (err, da
-            } else {
-                // tzid not present (calculate offset from original start)
-                newDate = new Date(date.setHours(date.getHours() - ((event.start.getTimezoneOffset() - date.getTimezoneOffset()) / 60)))
-            }
-            const start = moment(newDate)
-            console.log('Recurrence start:', start)
-        })
+    ```js
+    const localDay = new Intl.DateTimeFormat('de-CH', {
+        timeZone: event.start.tz ?? 'UTC',
+        year: 'numeric',
+        month: '2-digit',
+        day: '2-digit',
+    }).format(event.start);
+    ```
 
````
Done with option B: fe7769d109592e9e916dc9cc3a07401b8309028c.

KristjanESPERANTO commented on ical.js:
```diff
@@ -210,104 +142,96 @@ const dateParameter = function (name) {
+            return new Date(year, monthIndex, day, hour, minute, second);
           }
-        }
 
-        // Watch out for offset timezones
-        // If the conversion above didn't find any matching IANA tz
-        // And offset is still present
-        if (tz && tz.startsWith('(')) {
-          // Extract just the offset
-          const regex = /[+|-]\d*:\d*/;
-          offset = tz.match(regex);
-          found = offset;
-          tz = null;
-        }
+          const tzInfo = tzUtil.resolveTZID(normalizedTzId);
+          return tzInfo.iana
+            ? tzUtil.parseDateTimeInZone(value, tzInfo.iana)
+            : new Date(year, monthIndex, day, hour, minute, second);
+        };
```
Done with f2082fa4cdb975f6ee39b89e4e5c8313f2fdf9cd.

KristjanESPERANTO commented on examples/example-rrule-luxon.js:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  // Expand RRULE start dates within the range, keying each occurrence by its exact start time.
+  const instanceDates = new Map();
+  for (const date of event.rrule.between(rangeStart.toJSDate(), rangeEnd.toJSDate(), true)) {
+    const occurrence = DateTime.fromJSDate(date);
+    const iso = occurrence.toISO();
+    const lookupKey = iso.slice(0, 10);
+    if (event.recurrences && event.recurrences[lookupKey]) {
```
Done with a44f371a70ac2b995c9b5a10507b6f951beda2b6.

KristjanESPERANTO commented:
I’ve worked through CodeRabbit’s nitpicks and follow-up comments again :exploding_head:  The effort definitely exceeds the original scope (replace `moment-timezone`)  - but the feedback is valuable and the resulting cleanups substantially improve the codebase.

That said, the stricter behaviour means that consumers who relied on previous exceptions or leniency may experience issues. That should be highlighted clearly in the next release.

jens-maus commented:
> That said, the stricter behaviour means that consumers who relied on previous exceptions or leniency may experience issues. That should be highlighted clearly in the next release.

Thanks @KristjanESPERANTO for your efforts in this PR. This is highly appreciated. Can you, however, please quickly elaborate which exceptions or changes users might have to expect in which situations. The test runs look fine of course, but they might not catch all situations of course. So perhaps you might be able to outline the expected changes a bit so that I can add some extra breaking change comment (or something like that) to the next upcoming release. And please give me some time to do a brief review of your changes, this will certainly take some time...

KristjanESPERANTO committed:
docs: clarify DST handling in `convergeLocalInstant`

KristjanESPERANTO commented:
> Can you, however, please quickly elaborate which exceptions or changes users might have to expect in which situations.

I assume that most applications won’t notice these changes. The adjustments mostly surface when feeds have malformed offsets, rely on Windows display-name TZIDs, or consume the optional metadata we now preserve. If users already handled errors and treated all-day events carefully, they’re likely unaffected - but the cases where people did struggle should behave more predictably now.

However, here’s what users may encounter while upgrading:

1. **Malformed offsets now fail fast.**  
   _Before:_ Invalid TZID strings such as `"(UTC+Foo)"` slipped through and produced silently wrong `Date` values.  
   _After:_ We throw `TypeError: Invalid offset string`, so bad data surfaces immediately.  
   _Reason:_ Explicit failures make it easier to spot and fix broken feeds instead of propagating corrupt timestamps.  
   _Impact for upgraders:_ Wrap `ical.parse*` calls in try/catch (or keep existing error handling) so you can log/ignore bad events without crashing.

2. **Windows display names map to real IANA zones.**  
   _Before:_ Labels like `"(UTC+05:30) Chennai…"` remained as plain strings in `event.start.tz` and RRULE metadata.  
   _After:_ They resolve to canonical identifiers such as `Asia/Calcutta`.  
   _Reason:_ Using a single canonical name improves equality checks, makes snapshots predictable, and aligns with other timezone-aware libraries.  
   _Impact for upgraders:_ Update any code or fixtures that compare `tz` strings exactly; expect IANA names (or `Etc/GMT±X`) instead of raw Windows captions.

3. **All-day entries keep their wall-time metadata.**  
   _Before:_ `DTSTART;TZID=Europe/Berlin;VALUE=DATE:20240325` lost `dateOnly` and `tz`, encouraging consumers to assume midnight UTC.  
   _After:_ We still return the same UTC instant, but preserve `dateOnly === true` and the originating `tz`.  
   _Reason:_ Keeping wall-clock context prevents all-day events from drifting across dates when rendered in their home timezone.  
   _Impact for upgraders:_ Prefer deriving display dates via `event.start.tz` instead of assuming UTC midnight; all-day display logic becomes simpler and more accurate.

4. **Timezone metadata is non-enumerable.**  
   _Before:_ `Object.keys(event.start)` / `JSON.stringify(event.start)` exposed the `tz` property to every serialised payload.  
   _After:_ `tz` remains readable (`event.start.tz`) but is hidden from enumeration.  
   _Reason:_ This avoids leaking internal metadata into JSON without breaking existing code that relies on the property.  
   _Impact for upgraders:_ If you previously relied on `JSON.stringify` including `tz`, add it explicitly; otherwise nothing changes.

5. **RRULE string output looks slightly different.**  
   _Before:_ Even with a timezone, RRULEs often showed `DTSTART=…Z`, losing the original wall time.  
   _After:_ When possible we emit `DTSTART;TZID=Europe/Berlin:…`.  
   _Reason:_ RFC 5545 expresses DTSTART in local time when a TZID exists; keeping that form ensures recurrence expansion matches the source calendar.  
   _Impact for upgraders:_ Refresh any RRULE snapshots or tests that compare strings; consumers parsing the rule gain better DST fidelity without code changes.

6. **moment-timezone aliasing moved into `tz-utils`.**  
   _Before:_ Consumers had to call `moment.tz.link('Etc/Unknown|Etc/GMT')`, tying aliasing to the moment-timezone runtime.  
   _After:_ Aliases live in `require('node-ical/tz-utils').linkAlias('Etc/Unknown', 'Etc/GMT')`.  
   _Reason:_ Centralising alias support lets projects drop moment-timezone entirely while still configuring custom mappings.  
   _Impact for upgraders:_ Replace any `moment.tz.link` usage with `tz-utils.linkAlias`; you can remove `moment-timezone` from production dependencies if you no longer need it.


As a bonus idea for the next release or two: We could introduce built-in RRULE expansion support that leverages the new Intl helpers to handle recurring events more precisely – especially during DST transitions. This would help users who currently have to work manually with rrule.js. I just don't want to overload this PR any further.

KristjanESPERANTO committed:
fix: validate offset bounds in offsetLabelToMinutes

- Enforce IANA/ICS limits: minutes must be < 60, max absolute offset is ±14:00
- Prevents accepting impossible offsets like +12:90 or +23:59
- Returns undefined for invalid values

KristjanESPERANTO committed:
refactor(perf): cache Intl.DateTimeFormat instances in hot paths

- Add per-zone formatter cache (dtfCache) to reduce repeated Intl constructor churn
- Introduce getFormatter function to retrieve or create cached formatters
- Reuse cached formatters in formatDateForRrule and parseDateTimeInZone
- Cuts CPU and allocations during RRULE expansion and parsing

KristjanESPERANTO committed:
fix(tz): guard `parseDateTimeInZone` against invalid TZIDs

KristjanESPERANTO committed:
perf: memoize isValidIana checks

- Cache resolveZone normalization to avoid repeated Intl constructor throws
- Reuse validIanaCache to skip redundant exceptions during frequent resolveTZID calls
- Keeps the change minimal while improving hot-path performance

KristjanESPERANTO committed:
fix: preserve offset metadata in parseWithOffset

- Normalize offset strings with minutesToOffset before attaching tz metadata
- Return Dates via attachTz so downstream consumers can recover the original offset

KristjanESPERANTO committed:
fix: normalize Intl midnight rollover output

- document the pre-Node 22 Intl formatter bug and route midnight through normalizeMidnightParts
- expose the helper for tests and reuse it in formatDateForRrule/parseDateTimeInZone
- add a regression test covering the 24:00 formatter output

KristjanESPERANTO commented on tz-utils.js:
```diff
@@ -0,0 +1,628 @@
+      out[target] = Number(part.value);
+    }
+
+    if (out.hour === 24) {
+      out.hour = 0;
+    }
+
+    if (out.year && out.month && out.day && out.hour !== undefined && out.minute !== undefined && out.second !== undefined) {
+      return `${out.year}${pad2(out.month)}${pad2(out.day)}T${pad2(out.hour)}${pad2(out.minute)}${pad2(out.second)}`;
+    }
```
Done with 482d7caffe55bc42973d58d3fd031d1fd7be9fd7.

KristjanESPERANTO commented:
I have processed the last new comment from CodeRabbit. I am leaving the three new nitpicks as they are; I consider them not as crucial and would like to find an end for this PR. The last commits have not caused any changes to the above summary.

KristjanESPERANTO commented:
@jens-maus To reach a closure, I would stop here for now and wait for your feedback :slightly_smiling_face:

KristjanESPERANTO committed:
Merge branch 'master' into tz

KristjanESPERANTO commented:
@jens-maus, is there anything else I can do to help you move forward here? I know it's a huge piece of work, but it wouldn't be easy to split it into smaller PRs.

jens-maus commented:
Sorry @KristjanESPERANTO for the longer delay in reviewing this PR, but I was on vacation for some time and also have not had the opportunity yet to walk through your numerous changes.

I did now and tested it myself and I have to say they are of course quite comprehensive, but look perfectly fine so far. I indeed think it is time to merge this to the master branch and then publish a new 0.22.0 version which other users can then use and report any further issues. However, as we do have quite extensive test cases in our environment I also do not expect anything serious given that users are using node-ical through its official interface.

So thanks @KristjanESPERANTO for this great changes and all the work that you have put in! This is highly appreciated and I think this is a bit step forwarding in having a more smooth and streamlined node-ical experience and also given that moment-timezone is now removed as a dependency, this should also reduce the package size of node-ical considerably. So thanks again!

KristjanESPERANTO commented:
Thanks for checking and accepting! :rocket: 

Because I was interested in the numbers, I did some research:

| Metric (tarball contents, no deps)             | 0.21.0 (moment-timezone) | 0.22.0 (Intl utilities) | Net Change                 |
|------------------------------------------------|---------------------------|--------------------------|----------------------------|
| npm “Unpacked Size” (tarball contents)         | 78 kB                     | 98 kB                    | +20 kB  |
| Tarball size (compressed download)             | 23.4 kB                   | 29.4 kB                  | +6 kB                      |
| `node_modules` size after install              | ~9.2 MB                   | ~1.2 MB                  | **≈ 8 MB saved**           |
| Heap bump after `require('node-ical')`         | ~2.4 MB                   | ~0.85 MB                 | **≈ 1.5 MB less RAM**      |

So the slightly higher npm “Unpacked Size” is the cost of bundling our leaner Intl-based utilities, while users still benefit from much smaller installs and lower runtime memory.
